packages/trace/src/teleReceiver.ts

```diff
@@ -0,0 +1,435 @@
+    testResult.startTime = new Date(payload.startTime);
+    this._reporter.onTestBegin?.(test, testResult);
+  }
+
+  private _onTestEnd(testId: string, payload: JsonTestResultEnd) {
+    const test = this._tests.get(testId)!;
+    const result = test.resultsMap.get(payload.id)!;
+    result.duration = payload.duration;
+    result.status = payload.status;
+    result.errors = payload.errors;
+    result.attachments = payload.attachments;
+    this._reporter.onTestEnd?.(test, result);
+  }
+
+  private _onStepBegin(testId: string, resultId: string, payload: JsonTestStepStart) {
+    const test = this._tests.get(testId)!;
+    const result = test.resultsMap.get(resultId)!;
+    const step: TestStep = {
+      titlePath: () => [],
+      title: payload.title,
+      category: payload.category,
+      location: payload.location,
+      startTime: new Date(payload.startTime),
+      duration: 0,
+      steps: [],
+    };
+    (step as any).id = payload.id;
+    // TODO: implement nested steps.
+    result.stepMap.set(payload.id, step);
+    result.stepStack[result.stepStack.length - 1].steps.push(step);
+    result.stepStack.push(step);
+    this._reporter.onStepBegin?.(test, result, step);
+  }
+
+  private _onStepEnd(testId: string, resultId: string, payload: JsonTestStepEnd) {
+    const test = this._tests.get(testId)!;
+    const result = test.resultsMap.get(resultId)!;
+    const step = result.stepMap.get(payload.id)!;
+    const i = result.stepStack.indexOf(step);
+    if (i !== -1)
+      result.stepStack.splice(i, 1);
+    step.duration = payload.duration;
+    step.error = payload.error;
+    this._reporter.onStepEnd?.(test, result, step);
+  }
+
+  private _onError(error: TestError) {
+    this._reporter.onError?.(error);
+  }
+
+  private _onStdIO(type: 'stdout' | 'stderr', testId: string | undefined, resultId: string | undefined, data: string, isBase64: boolean) {
+    const chunk = isBase64 ? Buffer.from(data, 'base64') : data;
+    const test = testId ? this._tests.get(testId) : undefined;
+    const result = test && resultId ? test.resultsMap.get(resultId) : undefined;
+    if (type === 'stdout')
+      this._reporter.onStdOut?.(chunk, test, result);
+    else
+      this._reporter.onStdErr?.(chunk, test, result);
+  }
+
+  private _onEnd(result: FullResult) {
+    this._reporter.onEnd?.(result);
+  }
+
+  private _parseConfig(config: JsonConfig): FullConfig {
+    const fullConfig = baseFullConfig;
+    fullConfig.rootDir = config.rootDir;
+    fullConfig.configFile = config.configFile;
+    return fullConfig;
+  }
+
+  private _parseProject(project: JsonProject): TeleFullProject {
+    return {
+      id: project.id,
+      metadata: project.metadata,
+      name: project.name,
+      outputDir: project.outputDir,
+      repeatEach: project.repeatEach,
+      retries: project.retries,
+      testDir: project.testDir,
+      testIgnore: project.testIgnore,
+      testMatch: project.testMatch,
```

Comment:
I think `serializePatterns` turns regex into a string, so we should parse back here.